logicalplan/histogram_stats.go

```diff
@@ -0,0 +1,36 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
+package logicalplan
+
+import (
+	"github.com/prometheus/prometheus/util/annotations"
+
+	"github.com/thanos-io/promql-engine/query"
+)
+
+type DetectHistogramStatsOptimizer struct{}
+
+func (d DetectHistogramStatsOptimizer) Optimize(plan Node, _ *query.Options) (Node, annotations.Annotations) {
+	return d.optimize(plan, false)
+}
+
+func (d DetectHistogramStatsOptimizer) optimize(plan Node, decodeStats bool) (Node, annotations.Annotations) {
+	var stop bool
+	Traverse(&plan, func(node *Node) {
+		if stop {
+			return
+		}
+		switch n := (*node).(type) {
+		case *VectorSelector:
+			n.DecodeNativeHistogramStats = decodeStats
+		case *FunctionCall:
+			if n.Func.Name == "histogram_count" || n.Func.Name == "histogram_sum" {
```

Comment:
i think we never set the flag to false if we hit `histogram_fraction` under a `histogram_count` node

Reply:
Yes, but I think this query will return an empty result anyway because `histogram_fraction` expects a vector of histograms as input, and `histogram_count` produces a vector of floats.

Reply:
Imagine `histogram_count(histogram_fraction(H1) * H2)` kinda query! that would break right?

Reply:
In this case `histogram_fraction` also returns float as result, and histogram_count works only on histograms. But maybe it's better to add this reset to stay on the safe side. Would be nice if we had some type checking built in though.

Reply:
but if H2 is a histogram, wont "float*H2" also be a histogram?

Reply:
Ah right, but in the end we just want the count and right? So the multiplication will scale the count from H2 and in the end we will ignore the buckets.

Reply:
But probably sum could change!

Reply:
Let me add a test case for this.

Reply:
You are right, this would lead to wrong results. I added test cases and fixed the bug.